sycl/doc/extensions/proposed/sycl_ext_oneapi_joint_for.asciidoc

```diff
@@ -0,0 +1,210 @@
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+A very common idiom in SPMD programming sees developers distribute a range over
+the members of a group, by writing a loop that looks like this:
+
+[source,c++]
+----
+for (int i = it.get_local_id(0); i < N; i += it.get_local_range(0)) { ... }
+----
+
+The combination of setting the initial value based on the work-item index and
+setting the stride based on the number of work-items in the group ensures that
+any data accesses within the loop are appropriately interleaved to maximize
+performance.
+
+Although common, this idiom has several drawbacks: it is difficult to teach to
+developers unfamiliar with SPMD programming; it is easy to use incorrectly
+(e.g., by forgetting to set a different initial value on each work-item); and
+it is difficult to use generically (i.e., for different groups of work-items).
+
+This extension proposes to encapsulate this pattern in group algorithms, to
+simplify generic usage across different group types.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_JOINT_FOR` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Group algorithms
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template <typename Group, typename InputIterator, typename Function>
+Function joint_for_each(Group g, InputIterator first, InputIterator last, Function f);
+
+} // namespace sycl::ext::oneapi
+----
+
+_Constraints_: Available only if `sycl::is_group_v<std::decay_t<Group>>` is
+true and `InputIterator` is a random access iterator.
+
+_Preconditions_: `first` and `last` must be the same for all work-items in
+group `g`, and `f` must be an immutable callable with the same type and state
+for all work-items in group `g`.
+
+_Effects_: Blocks until all work-items in group `g` have
+reached this synchronization point, then applies `f` to the result of
+dereferencing every iterator in the range [`first`, `last`).
+The range is distributed across the work-items in group `g`.
+
+_Synchronization_: The call to this function in each work-item
+happens before the algorithm begins execution.
+The completion of the algorithm happens before any work-item
+blocking on the same synchronization point is unblocked.
+
+_Remarks_: If `f` returns a result, the result is ignored.
+If the range [`first`, `last`) is not evenly divisible by the number of
+work-items in group `g`, the number of times that `f` is invoked will differ
+across work-items.
+
+_Returns_: `f`.
```

Comment:
Why is `f` returned?

It might help to add an example at the end showing why this is useful.

Reply:
Again, this was for alignment with `std::for_each`.  I have no idea why it works this way, but it does!

Reply:
The overload of `std::for_each` that takes an execution policy does not return `f` and has this explanation:

> [Note [4]: Does not return a copy of its Function parameter, since parallelization often does not permit efficient state accumulation. — end note]

I think the overload that does *not* take an execution policy allows `f` to be mutable.  That function returns `f` because it represents the final state of the function object after calling it for each item.

Since your extension requires `f` to be immutable, I think it is closer to the C++ overload that takes an execution policy, so I think yours should return `void`.

Reply:
Oh, I missed that!  I agree that we should match the execution policy overload.  Fixed in 34a6d888661a2e8b444d73484376958cc2a617e2.